common/src/main/java/com/amazonaws/services/schemaregistry/common/SchemaByDefinitionFetcher.java

```diff
@@ -0,0 +1,114 @@
+package com.amazonaws.services.schemaregistry.common;
+
+import com.amazonaws.services.schemaregistry.common.configs.GlueSchemaRegistryConfiguration;
+import com.amazonaws.services.schemaregistry.exception.AWSSchemaRegistryException;
+import com.amazonaws.services.schemaregistry.utils.AWSSchemaRegistryConstants;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.SneakyThrows;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Fetches the schema version for the given schema definition optionally registering the schema if required.
+ */
+public class SchemaByDefinitionFetcher {
+    @NonNull
+    private final AWSSchemaRegistryClient awsSchemaRegistryClient;
+
+    @NonNull
+    private final GlueSchemaRegistryConfiguration glueSchemaRegistryConfiguration;
+
+    @NonNull
+    @VisibleForTesting
+    protected final LoadingCache<Schema, UUID> schemaDefinitionToVersionCache;
+
+    public SchemaByDefinitionFetcher(
+        final AWSSchemaRegistryClient awsSchemaRegistryClient,
+        final GlueSchemaRegistryConfiguration glueSchemaRegistryConfiguration) {
+        this.awsSchemaRegistryClient = awsSchemaRegistryClient;
+        this.glueSchemaRegistryConfiguration = glueSchemaRegistryConfiguration;
+
+        this.schemaDefinitionToVersionCache = CacheBuilder.newBuilder()
+            .maximumSize(glueSchemaRegistryConfiguration.getCacheSize())
+            .refreshAfterWrite(glueSchemaRegistryConfiguration.getTimeToLiveMillis(), TimeUnit.MILLISECONDS)
+            .build(new SchemaDefinitionToVersionCache());
+    }
+
+    /**
+     * Get Schema Version ID by following below steps :
+     * <p>
+     * 1) If schema version id exists in registry then get it from registry
+     * 2) If schema version id does not exist in registry
+     * then if auto registration is enabled
+     * then if schema exists but version doesn't exist
+     * then
+     * 2.1) Register schema version
+     * else if schema does not exist
+     * then
+     * 2.2) create schema and register schema version
+     *
+     * @param schemaDefinition Schema Definition
+     * @param schemaName       Schema Name
+     * @param dataFormat       Data Format
+     * @param metadata         metadata for schema version
+     * @return Schema Version ID
+     * @throws AWSSchemaRegistryException on any error while fetching the schema version ID
+     */
+    @SneakyThrows
+    public UUID getORRegisterSchemaVersionId(
```

Comment:
Whats the benefit of changing to this?

Reply:
If you meant, refactoring the class, it's for separating out the auto-registration logic so that the results can be cached. It wasn't possible earlier as this logic was part of AWSSchemaRegistryClient which is common for serializer and de-serializer. Adding a cache in AWSSchemaRegistryClient will instantiate it for both the serializer and de-serializer which is not optimal.